src/benchmarks/micro/libraries/System.Diagnostics/Perf_Activity.cs

```diff
@@ -0,0 +1,373 @@
+
+            Enumerator<IEnumerable<ActivityLink>, ActivityLink, int>.AllocationFreeForEachDelegate foreachDelegate
+                = Enumerator<IEnumerable<ActivityLink>, ActivityLink, int>.FindAllocationFreeForEach(s_Activity.Links);
+
+            for (int i = 0; i < NumberOfActivities; i++)
+            {
+                foreachDelegate(s_Activity.Links, ref total, ForeachLinkRef);
+            }
+        }
+
+        private static readonly Enumerator<IEnumerable<ActivityLink>, ActivityLink, int>.ForEachDelegate ForeachLinkRef = ForeachLink;
+
+        private static bool ForeachLink(ref int total, ActivityLink link)
+        {
+            total++;
+            return true;
+        }
+
+        [Benchmark]
+        public void EnumerateActivityEvents()
+        {
+            int total = 0;
+
+            for (int i = 0; i < NumberOfActivities; i++)
+            {
+                foreach (var @event in s_Activity.Events)
+                {
+                    total++;
+                }
+            }
+        }
+
+        [Benchmark]
+        public void OTelHelperActivityEvents()
+        {
+            int total = 0;
+
+            Enumerator<IEnumerable<ActivityEvent>, ActivityEvent, int>.AllocationFreeForEachDelegate foreachDelegate
+                = Enumerator<IEnumerable<ActivityEvent>, ActivityEvent, int>.FindAllocationFreeForEach(s_Activity.Events);
+
+            for (int i = 0; i < NumberOfActivities; i++)
+            {
+                foreachDelegate(s_Activity.Events, ref total, ForeachEventRef);
+            }
+        }
+
+        private static readonly Enumerator<IEnumerable<ActivityEvent>, ActivityEvent, int>.ForEachDelegate ForeachEventRef = ForeachEvent;
+
+        private static bool ForeachEvent(ref int total, ActivityEvent @event)
+        {
+            total++;
+            return true;
+        }
+
+        [Benchmark]
+        public void EnumerateActivityLinkTags()
+        {
+            int total = 0;
+
+            for (int i = 0; i < NumberOfActivities; i++)
+            {
+                foreach (var tag in s_ActivityLink.Tags)
+                {
+                    total++;
+                }
+            }
+        }
+
+        [Benchmark]
+        public void OTelHelperActivityLinkTags()
+        {
+            int total = 0;
+
+            Enumerator<IEnumerable<KeyValuePair<string, object>>, KeyValuePair<string, object>, int>.AllocationFreeForEachDelegate foreachDelegate
+                = Enumerator<IEnumerable<KeyValuePair<string, object>>, KeyValuePair<string, object>, int>.FindAllocationFreeForEach(s_ActivityLink.Tags);
+
+            for (int i = 0; i < NumberOfActivities; i++)
+            {
+                foreachDelegate(s_ActivityLink.Tags, ref total, ForeachTagObjectRef);
+            }
+        }
+
+        // A helper class for enumerating over IEnumerable<TItem> without allocation if a struct enumerator is available.
+        private class Enumerator<TEnumerable, TItem, TState>
```

Comment:
>Enumerator [](start = 22, length = 10)

I am not really sure if it is good to test the internals with the reflection and code emit. what is really the goal here?

Reply:
These benchmarks were written for the other PR, really. I needed to make sure the struct enumerators worked for completely removing the allocation with the reflection mechanism, and also show there was an allocation reduction for the standard enumeration too. I opened this PR because I think @noahfalk said it would be useful. So my goal now is just to provide usefulness to future users of Activity 😄 I can remove the OpenTelemetry stuff, or I can close the PR completely if we don't think it will be that useful. LMK

Reply:
Adding perf tests for Activity is useful. I suggest removing all OTel stuff and keep the simple test cases for the enumerating tags, links...etc. Thanks a lot for all your provided help.